design-documents/graph-ql/coverage/reward-points.graphqls

```diff
@@ -0,0 +1,70 @@
+type Customer {
+    reward_points: RewardPoints @doc(description: "Customer reward points details")
+}
+
+type Cart {
+    applied_reward_points: RewardPointsAmount @doc(description: "Reward points ammount applied to the cart")
+}
+
+type Mutation {
+    applyRewardPointsToCart(cartId: ID!): ApplyRewardPointsToCartOutput @doc(description: "Apply all available points up to the cart total. Partial redemption is not available")
+    removeRewardPointsFromCart(cartId: ID!): RemoveRewardPointsFromCartOutput @doc(description: "Cancel usage of the previously applied reward points to cart")
+}
+
+type ApplyRewardPointsToCartOutput {
+    cart: Cart
```

Comment:
We've discussed that, for object types, we want non-scalar fields to be nullable, which we've followed here.

But, for mutation return types, I'm wondering if we expect a case where we could do a _write_ to the cart service, but then not be able to do a read?

The reason for nullable on non-scalar types is to guard ourselves from cascading failures when some services are down. Do we have concerns that cross service boundaries here (honest question, I truly do not know)?

One trade-off we make with nullability is that the client-side has to do conditional checks when it gets a response, and handle additional states.

```js
function MyComponent() {
   // ... run mutation here
   if (response.error) {
      // handle error updating cart
   }
   if (response.cart) {
      // safe to use cart data when rendering
   } else {
      // handle case where we were able to successfully update the cart,
      // but weren't able to query for new results
   }
}
```

when we go non-nullable, that can turn into:

```js
function MyComponent() {
   if (response.error) {
      // handle error updating cart
   }
   
   // render UI with cart data
}
```

I think non-nullable makes a bit more sense in the context of a mutation return type since we have more context about what the client would expect. Thoughts?

Reply:
Sounds reasonable, fixing it here. @DrewML , can I ask you to write these rules down somewhere in https://github.com/magento/architecture/tree/master/design-documents/graph-ql ?

1. Non-nullable for mutations
1. Nullable for objects in queries
1. Non-nullable for scalars in queries

Reply:
Thanks for updating!

Left myself a TODO to add a document covering our experiences/preferences with nullability